### PR TITLE
add tarwriter

### PR DESCRIPTION
**Pull request by tonistiigi**
Add tarwriter and refactor `FS` so that memory-only subdirs can be exported with the tarball.

@AkihiroSuda

### CODE REVIEW

File: fs.go [hunk 1]
```diff
@@ -37,36 +39,80 @@ func (fs *fs) Open(p string) (io.ReadCloser, error) {
...
 }
 
 func (fs *subDirFS) Walk(ctx context.Context, fn filepath.WalkFunc) error {
-	main := &StatInfo{Stat: &fs.stat}
-	if !main.IsDir() {
-		return errors.Errorf("fs subdir not mode directory")
+	for _, d := range fs.dirs {
+		fi := &StatInfo{Stat: &d.Stat}
+		if !fi.IsDir() {
+			return errors.Errorf("fs subdir not mode directory")
```
Comment (AkihiroSuda) [inline fs.go]: errorf d.Stat.Path?

File: fs.go [hunk 2]
```diff
@@ -37,36 +39,80 @@ func (fs *fs) Open(p string) (io.ReadCloser, error) {
...
+		}
+		if err := d.FS.Walk(ctx, func(p string, fi os.FileInfo, err error) error {
+			stat, ok := fi.Sys().(*types.Stat)
+			if !ok {
+				return errors.Wrapf(err, "invalid fileinfo without stat info: %s", p)
+			}
+			stat.Path = path.Join(d.Stat.Path, stat.Path)
+			if stat.Linkname != "" {
+				if fi.Mode()&os.ModeSymlink != 0 {
+					if strings.HasPrefix(stat.Linkname, "/") {
```
Comment (AkihiroSuda) [inline fs.go]: can we move this to `_unix.go`?

Comment (tonistiigi) [reply fs.go]: What part specifically are you concerned about? `types.Stat` is supposed to be cross platform and always normalized to unix slashes. (unlike syscall.Stat_t)

File: fs.go (hunk 1 above)
Comment (tonistiigi) [reply fs.go]: updated

**Pull request merged**